### PR TITLE
BUG: to_sql convert index name to string (#15404)

### LINKED ISSUE #15404: BUG: to_sql errors with numeric index name - needs conversion to string
#### Code Sample, a copy-pastable example if possible

```python

import pandas as pd
import sqlalchemy as sa


d = {'2000': 1, '2001': 2, '2002': 3, '2003': 4}

df = pd.DataFrame(list(d.items()))
df = df.set_index([0])

db = sa.create_engine("sqlite:///test.db")
df.to_sql("test_tbl", db, if_exists="append")


```
#### Problem description

This errors as follows:

```
  File "sql.py", line 12, in <module>
    df.to_sql("test_tbl", db, if_exists="append")
  File "/home/zzz/.virtualenvs/sqlalchemy/lib/python3.5/site-packages/pandas/core/generic.py", line 1201, in to_sql
    chunksize=chunksize, dtype=dtype)
  File "/home/zzz/.virtualenvs/sqlalchemy/lib/python3.5/site-packages/pandas/io/sql.py", line 470, in to_sql
    chunksize=chunksize, dtype=dtype)
  File "/home/zzz/.virtualenvs/sqlalchemy/lib/python3.5/site-packages/pandas/io/sql.py", line 1146, in to_sql
    schema=schema, dtype=dtype)
  File "/home/zzz/.virtualenvs/sqlalchemy/lib/python3.5/site-packages/pandas/io/sql.py", line 563, in __init__
    self.table = self._create_table_setup()
  File "/home/zzz/.virtualenvs/sqlalchemy/lib/python3.5/site-packages/pandas/io/sql.py", line 770, in _create_table_setup
    for name, typ, is_index in column_names_and_types]
  File "/home/zzz/.virtualenvs/sqlalchemy/lib/python3.5/site-packages/pandas/io/sql.py", line 770, in <listcomp>
    for name, typ, is_index in column_names_and_types]
  File "/home/zzz/.virtualenvs/sqlalchemy/lib/python3.5/site-packages/sqlalchemy/sql/schema.py", line 1245, in __init__
    self._init_items(*args)
  File "/home/zzz/.virtualenvs/sqlalchemy/lib/python3.5/site-packages/sqlalchemy/sql/schema.py", line 79, in _init_items
    item._set_parent_with_dispatch(self)
AttributeError: 'int' object has no attribute '_set_parent_with_dispatch'
```

Commenting out the df.set_index line results in expected behaviour though with a simple integer index. Using sqlite3 as the library also results in errors.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-62-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8

pandas: 0.19.2
nose: None
pip: 9.0.1
setuptools: 34.1.1
Cython: None
numpy: 1.12.0
scipy: None
statsmodels: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.1.5
pymysql: None
psycopg2: None
jinja2: None
boto: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by redbullpeter**
…rs with numeric index name - needs conversion to string

 - [x] closes #15404
 - [x] tests passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@redbullpeter Thanks

Can you add a test case for this? (in pandas/tests.io/test_sql.py) You can put the test in the `_TestSQLApi` class

**Comment by redbullpeter** [comment]
While writing the test cases I discovered I jumped the gun on this. Will probably have to rescind the pull request.

**Comment by redbullpeter** [comment]
OK done. For a while there it seemed like I was going down a rabbit hole that got more confusing as I picked at it.

File: pandas/io/sql.py [hunk 1]
```diff
@@ -750,7 +750,8 @@ def _get_column_names_and_types(self, dtype_mapper):
             for i, idx_label in enumerate(self.index):
                 idx_type = dtype_mapper(
                     self.frame.index.get_level_values(i))
-                column_names_and_types.append((idx_label, idx_type, True))
+                column_names_and_types.append(((text_type(idx_label)),
```
Comment (jorisvandenbossche) [inline pandas/io/sql.py]: minor issue, but I think you have here one pair of brackets that is not needed

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good!

Only very minor comment

File: pandas/io/sql.py (hunk 1 above)
Comment (redbullpeter) [inline pandas/io/sql.py]: Three lines further down the same kind of type_text conversion was done with a set of extra brackets. I just copied the style for consistency.

Comment (jorisvandenbossche) [inline pandas/io/sql.py]: then you can remove it in both places. This is just redundant in python syntax, and not enhancing readability IMO (which could be a reason to have redundant chars)

Comment (redbullpeter) [inline pandas/io/sql.py]: Removed it from my changes. Left the other one alone as on closer inspection the brackets were not extraneous after all, the line breaks just made it hard to see the pairing.

**Comment by jorisvandenbossche** [comment]
@redbullpeter Thanks a lot!

**Comment by jorisvandenbossche** [comment]
@redbullpeter Whoops, forgot to ask you for a whatsnew note, so quickly added it: https://github.com/pandas-dev/pandas/commit/54b6c6e1c443b992a1df3443669a59dbe430271f

**Pull request merged**